### ISSUE TITLE
Webpack build hangs with parallel: true

### ISSUE THREAD

**Issue by Pajn** [webpack-contrib/terser-webpack-plugin#21]
* Operating System: Windows 10 1810, WSL, Ubuntu 18.04
* Node Version: 8.12.0
* NPM Version: 6.4.1
* webpack Version: 4.19.1
* terser-webpack-plugin Version: 1.1.0

### Expected Behavior

That the webpack build completes

### Actual Behavior

The webpack build hangs forever (or at least 30 minutes which is the longest I have lasted) after outputting
```
Compilation  starting…
```

However changing `parallel: true` to `parallel: false` in terser options will cause the build to complete in normal time (around 40s)

### Code

```js
new TerserPlugin({
        terserOptions: {
          parse: {
            // we want terser to parse ecma 8 code. However, we don't want it
            // to apply any minfication steps that turns valid ecma 5 code
            // into invalid ecma 5 code. This is why the 'compress' and 'output'
            // sections only apply transformations that are ecma 5 safe
            // https://github.com/facebook/create-react-app/pull/4234
            ecma: 8
          },
          compress: {
            ecma: 5,
            warnings: false,
            // Disabled because of an issue with Uglify breaking seemingly valid code:
            // https://github.com/facebook/create-react-app/issues/2376
            // Pending further investigation:
            // https://github.com/mishoo/UglifyJS2/issues/2011
            comparisons: false,
            // Disabled because of an issue with Terser breaking valid code:
            // https://github.com/facebook/create-react-app/issues/5250
            // Pending futher investigation:
            // https://github.com/terser-js/terser/issues/120
            inline: 2
          },
          mangle: {
            safari10: true
          },
          output: {
            ecma: 5,
            comments: false,
            // Turned on because emoji and regex is not minified properly using default
            // https://github.com/facebook/create-react-app/issues/2488
            ascii_only: true
          }
        },
        // Use multi-process parallel running to improve the build speed
        // Default number of concurrent runs: os.cpus().length - 1
        parallel: true,
        // Enable file caching
        cache: true,
        sourceMap: shouldUseSourceMap
      })
```

Full webpack config: https://github.com/Pajn/tscomp/blob/cra2/config/webpack.config.prod.js

### How Do We Reproduce?

I suspect this can be related to WSL. If you have some debugging tips I can probably run them.

For the exact same state as I have:
1. Clone https://github.com/Pajn/tscomp
2. `git checkout terser-bug`
2. Clone https://github.com/Pajn/RAXA
3. `cd RAXA/packages/web`
4. `yarn`
5. `yarn add --dev path/to/cloned/tscomp/repo`
6. `yarn build`, or for more information `NODE_ENV=production webpack --config node_modules/tscomp/config/webpack.config.prod.js --verbose --info-verbosity verbose --display verbose --profile`

**Comment by Pajn** [comment]
To set up WSL you need Windows 10, then you should be able to follow this guide https://docs.microsoft.com/en-us/windows/wsl/install-win10

After that you can basically treat is as a normal Ubuntu Linux so install node by following this https://nodejs.org/en/download/package-manager/#debian-and-ubuntu-based-linux-distributions
Git can be installed via `sudo apt install git`

The only consideration with a case like this is that Linux files are not visible to Windows, so to be able to open files in a graphical editor you need to save them under `/mnt/c`

**Comment by evilebottnawi** [comment]
@Pajn oh, i don't have `windows` 10 right now, can debug this only on next week, feel free to debug problem self. Here default arguments to `work-farm` https://github.com/webpack-contrib/terser-webpack-plugin/blob/master/src/TaskRunner.js#L36, maybe we need improve this check `process.platform === 'win32'` and pass `maxConcurrentCallsPerWorker: 1` on WSL

**Comment by Pajn** [comment]
You are right that worker-farm is somehow related.
This is what I get from pstree in a stuck build
```
  │   │   ├─node /usr/share/yarn/bin/yarn.js build
  │   │   │   ├─sh -c tscomp build
  │   │   │   │   └─node /c/Users/Rasmus/Development/RAXA/packages/web/node_modules/.bin/tscomp build
  │   │   │   │       ├─node /c/Users/Rasmus/Development/RAXA/packages/web/node_modules/tscomp/scripts/build.js
  │   │   │   │       │   ├─node /c/Users/Rasmus/Development/RAXA/packages/web/node_modules/tscomp/node_modules/worker-farm/lib/child/index.js /usr/bin/node /c/Users/Rasmus/Development/RAXA/packages/web/node_modules/tscomp/scripts/build.js
  │   │   │   │       │   │   └─5*[{node}]
  │   │   │   │       │   ├─node /c/Users/Rasmus/Development/RAXA/packages/web/node_modules/tscomp/node_modules/worker-farm/lib/child/index.js /usr/bin/node /c/Users/Rasmus/Development/RAXA/packages/web/node_modules/tscomp/scripts/build.js
  │   │   │   │       │   │   └─5*[{node}]
  │   │   │   │       │   └─9*[{node}]
  │   │   │   │       └─5*[{node}]
  │   │   │   └─9*[{node}]
```

Changing that condition so that `maxConcurrentCallsPerWorker: 1` is passed (process.platform is linux) did not help unfortunately.

I tried running the basic example from the worker-farm readme and that worked without hanging so I guess there some problem with file locking.

I'll continue the debugging tomorrow.

**Comment by Pajn** [comment]
I have four, however the above pstree output is from a run where I limited the number of workers to two, just to see if it did help.

I modified the code so that it would still use `worker-farm` with only one worker, and then it did finish again. I also tried to attach strace to see if I got some hint of the problem but did not find anything useful in the output. If you are interested I can of course share those outputs.

**Comment by evilebottnawi** [comment]
@Pajn Feel free to publish outputs

**Comment by evilebottnawi** [comment]
Guys i try to reproducible problem but i can't, maybe can create really minimum reproducible test repo with really minimum setup?

**Comment by charsleysa** [comment]
@evilebottnawi you should be able to recreate this by compiling the an Angular demo app (any should do) with `aot` and `optimization` flags turned on.

I get this issue where the TerserPlugin seems to hang with a lot of node processes showing in Task Manager.

**Comment by evilebottnawi** [comment]
@charsleysa can you provide instruction for minimum reproducible test repo?

**Comment by charsleysa** [comment]
@evilebottnawi it seems to be something to do with having a project sufficiently large enough producing source maps on a production build.

The following instructions assume you have already setup WSL with Node 10.x (node inside WSL) using standard setup instructions. The below instructions are executed inside bash.

1) clone this repo https://github.com/formio/angular-demo
2) cd into the cloned repo directory
3) install the angular cli `npm install -g @angular/cli`
4) install the repo requirements `npm install`
5) run a standard build just to check that everything is working `ng build`
6) run a production build with source maps turned on `ng build --configuration=production --project=demo --source-map=true`
7) if you watch Task Manager you'll notice that eventually the node processes will just cease utilizing CPU resource with the build reporting `92% chunk asset optimization TerserPlugin` (as pictured below)

![image](https://user-images.githubusercontent.com/1219661/51524203-6af4d700-1e92-11e9-8180-71209b7c5dae.png)

**Comment by evilebottnawi** [comment]
@charsleysa thanks for instructions, anyway feel free to investigate problem(s), i think problem inside `worker-farm`, maybe you can find problem faster than i can, because i need time to setup WSL

**Comment by andreialecu** [comment]
WSL is becoming more and more popular. I'm seeing the same problem as well and can confirm that `parallel: false` works. If #6 is applied to make parallel the default, then this needs to be fixed first.

**Comment by evilebottnawi** [comment]
@andreialecu can you investigate why it is happens? Don't have WSL around and don't have windows machine near, maybe you can find problem

**Comment by andreialecu** [comment]
I've been digging at it and from my initial findings it seems like this callback is never ran:

https://github.com/webpack-contrib/terser-webpack-plugin/blob/master/src/TaskRunner.js#L74

However, the task itself and the code in `minify.js` is called, and is returning properly. I'm trying to investigate why it never gets there now. I suspected it would be a problem in `worker-farm` but the examples in the `worker-farm` repo and all of its tests pass properly.

**Comment by evilebottnawi** [comment]
@andreialecu yep, problem in `worker-farm`, maybe we need to setup more options (https://github.com/webpack-contrib/terser-webpack-plugin/blob/master/src/TaskRunner.js#L34) for WSL

**Comment by evilebottnawi** [comment]
What is `process.platform` on WSL?

**Comment by andreialecu** [comment]
But like I said, `worker-farm`'s tests entire test suite passes properly.

**Comment by andreialecu** [comment]
Further progress on debugging this, it seems like the forked process' `on('message')` handler in `worker-farm` is never called when running with this plugin on WSL:

https://github.com/rvagg/node-worker-farm/blob/master/lib/farm.js#L145

However, when running `worker-farm` outside of this scenario, it **is** called. I verified using this: https://github.com/rvagg/node-worker-farm/tree/master/examples/basic

What could be different?

**Comment by andreialecu** [comment]
I think I found the root cause. 

It is related to the length of the result being passed from the worker process back to the farm via `process.send()`. If I make it send a simple `'hello'` instead of the entire minified script result it seems that all callbacks are called.

Making this callback with a shorter string results in all callbacks working:
https://github.com/webpack-contrib/terser-webpack-plugin/blob/master/src/worker.js#L17

I'm trying to reproduce it here https://github.com/rvagg/node-worker-farm/blob/master/examples/basic/index.js by changing the code to send a 10MB result and it works perfectly:

```javascript
module.exports = function (inp, callback) {
  const msg = 'a'.repeat(10 * 1000 * 1000);
  callback(null, inp + ' BAR (' + process.pid + ')' + msg)
}
```

So I'm not sure what is going on.

**Comment by evilebottnawi** [comment]
@Pajn Maybe some characters break this :confused:

**Comment by andreialecu** [comment]
@evilebottnawi Well, I just tested that by dumping the objects returned by the `minify()` worker functions to local `.json` files then trying to read each of them in simple workers in the basic example above, and then pass them on to the main process. Everything worked. 

I'm stumped.

Perhaps someone can take over. Running out of time assigned to this right now.

**Comment by evilebottnawi** [comment]
weird :confused:

**Comment by andreialecu** [comment]
A possible workaround for this would be to save all minified scripts to temp files from the worker and call back with the path to the temp files, then load them back in the master. This would probably fix it. How would you feel about this?

**Comment by evilebottnawi** [comment]
@andreialecu as workaround sounds good, but will be great find real problem why it is doesn't work as expected, maybe bug in `WSL` and we should report about this bug. Because on `macos`/`linux`/`windows` without WSL it is works as expected. So i think problem in `WSL`. If problem won't fix we implement workaround. Write files in temp directory decrease perf and it is not good.

**Comment by joewood** [comment]
When this works through a workaround, is the temp directory in **/tmp** or is it under **/mnt/c/** ? I'm wondering if this is a file system related bug.

**Comment by andreialecu** [comment]
On WSL: 

```
$ node -pe 'os.tmpdir()'
/tmp
```

**Comment by joewood** [comment]
I wonder if this fails under **/mnt/c/Users/[windowsuser]/temp** - which is (usually) NTFS.

**Comment by evilebottnawi** [comment]
@joewood maybe somebody has image for virtualbox with installed WSL?

**Comment by joewood** [comment]
I can try later today on my regular machine as I'm seeing this issue too. I suspect this is a WSL bug, the idea of trying on the different disk formats is to compare on the WSL formatted disk vs. regular NTFS (which has some differences).

**Comment by andreialecu** [comment]
The problem is with interprocess communication not with anything file system related. /tmp isn't used. It was just a suggestion I made to use as a workaround so it is just hypothetical at this point.

I could not reproduce the error outside of webpack so I'm still not sure what the root cause is.

**Comment by andreialecu** [comment]
Take a look here for Windows VMs: https://developer.microsoft.com/en-us/windows/downloads/virtual-machines

**Comment by evilebottnawi** [comment]
@andreialecu already work on this, i will put information here

**Comment by randak** [comment]
This thread caught my eye as I was searching for possible reasons we are seeing CI builds occasionally  time out (webpack hanging for 45+ minutes). We are not running WSL, but this in particular stood out to me:
> if you watch Task Manager you'll notice that eventually the node processes will just cease utilizing CPU resource with the build reporting 92% chunk asset optimization TerserPlugin (as pictured below)

I've seen our build hang for a while running locally on my Mac, and it was always at 92%. I haven't been able to get any more details out of our CI build hangs yet, but I'm going to test the `parallel` setting and see if it makes any difference. 

I should note that this is an intermittent issue for us; I'm not sure if other people are seeing this consistently or not.

**Comment by evilebottnawi** [comment]
Tested on latest version and all works fine, i think problem was fixed inside WSL

**Comment by evilebottnawi** [comment]
Try to reproduce with https://github.com/webpack-contrib/terser-webpack-plugin/issues/21#issuecomment-456324702

**Comment by evilebottnawi** [comment]
Parallels option works fine on latest WSL version

**Comment by andreialecu** [comment]
@evilebottnawi what Windows version are you testing on? 1903?

**Comment by evilebottnawi** [comment]
@andreialecu virtual machine from windows, how i can get version of windows?

**Comment by evilebottnawi** [comment]
I am trying to reproduce but can't, all works fine

**Comment by andreialecu** [comment]
Run `winver` in a command prompt or from run. Should say version 1809 or 1903.

**Comment by evilebottnawi** [comment]
@andreialecu Version 1803 (OS Build 17134.706)

**Comment by evilebottnawi** [comment]
@andreialecu maybe you can create minimum reproducible test repo?

**Comment by andreialecu** [comment]
I'm also on 1803, which is the currently widely available Windows version. 1903 is supposed to be released this month, and it has some WSL fixes, so maybe it might impact this issue. I haven't tested 1903 yet.

They also announced WSL2.0 which is a complete architecture change from WSL1.0 and whatever is causing this issue should be fixed by that, because it uses a real Linux kernel among other things. https://devblogs.microsoft.com/commandline/announcing-wsl-2/

That being said, I was just able to reproduce it on 1803 using https://github.com/webpack-contrib/terser-webpack-plugin/issues/21#issuecomment-456324702

One thing to check (just suggesting it, since it looks like you're not a Windows guy :) - are you sure you're running under WSL and not the regular Command Prompt?

**Comment by andreialecu** [comment]
Attaching repro proof:

![image](https://user-images.githubusercontent.com/697707/57703768-5fa80000-7669-11e9-9b5a-c66f36b5148c.png)

![image](https://user-images.githubusercontent.com/697707/57703840-7d756500-7669-11e9-9a7a-50784468c770.png)

**Comment by evilebottnawi** [comment]
@andreialecu yes

>  node -e 'console.log(require("os").release())'
> 4.4.0-17134-Microsoft

**Comment by evilebottnawi** [comment]
Investigate deeply, thanks for fast reply

**Comment by andreialecu** [comment]
@evilebottnawi actually, minor correction on my previous comment. I'm on 1809, not 1803. You're one version behind. If you definitely cannot reproduce it, that may point out why.

**Comment by evilebottnawi** [comment]
@andreialecu reproduce problem :star: WIP on fix

**Comment by endiliey** [comment]
I am on 1803. Surprisingly it worked now on WSL😃.Previously it wasn't working,. had to manually disable parallel

![image](https://user-images.githubusercontent.com/17883920/57705256-1621da00-7696-11e9-8b22-ad8556b1b411.png)

**Comment by evilebottnawi** [comment]
Repo https://github.com/webpack-contrib/terser-webpack-plugin/issues/21#issuecomment-456324702 is very hard :disappointed: i was spent around hour to search configuration and no luck, in future all big repositories will be ignored, it is hard to debug and it spends a lot of time

**Comment by endiliey** [comment]
I think https://github.com/webpack-contrib/terser-webpack-plugin/issues/21#issuecomment-456324702 webpack config is defined/ constructed dynamically on `@angular/cli` which is another huge project.

**Comment by evilebottnawi** [comment]
@endiliey @andreialecu looks it is not freeze, try to wait 5-25 minutes (depends on you hardware), problem what source map generation require many times

**Comment by evilebottnawi** [comment]
In webpack@5 we update https://github.com/mozilla/source-map to latest version (WASM) and it should solve problem with perf, but here we can't do this :disappointed:

**Comment by evilebottnawi** [comment]
Tried on many repos and all works fine, in some repos i have waited around 20 minutes due it is really big application. Just ensure it is not problem in plugin with WSL, just disable source map and try again

**Comment by andreialecu** [comment]
Did you see 0% cpu for node processes for a very long time while you waited, and it eventually finished? Did you try with `parallel: false`?

A problem still exists on WSL that is not present for native Windows or native Linux builds. If you go through my comments above you can use a debugger to attach and break on something inside the terser plugin code even if it is running via `ng build`.

Anyway, it's probably not worth working too much on this because I assume that it will automagically fix itself on WSL 2.0, and in the mean time `parallel: false` works without huge performance penalties.

**Comment by endiliey** [comment]
I have been running it for 2 hours but it still stucks in 92%.  Also 0% node process

**Comment by evilebottnawi** [comment]
@endiliey @andreialecu it is very strange, 20 min for https://github.com/webpack-contrib/terser-webpack-plugin/issues/21#issuecomment-456312880 and all were minified, maybe you can create other minimum reproducible test repo?

**Comment by evilebottnawi** [comment]
I am on SSD, maybe WSL has timeouts and in some cases it is failed :confused:

**Comment by evilebottnawi** [comment]
I can disable parallels on WSL by defaults, by want to ensure it is real problem in WSL, not in webpack or other ecosystem

**Comment by evilebottnawi** [comment]
Okey, we disable parallels on WSL and for 2.0 we migrate on `jest-worker` (they work with WSL) https://github.com/webpack-contrib/terser-webpack-plugin/issues/89

**Comment by evilebottnawi** [comment]
Somebody can test master version on WSL?

**Comment by evilebottnawi** [comment]
Can't setup azure pipelines for WSL, if anyone had experience please put here :smile:

**Comment by Pajn** [comment]
I'm sorry. I gave up on trying to use Windows for development because problems like this was the norm and not an exception so now I only have Linux to test on.

**Comment by evilebottnawi** [comment]
I am closing issue, if somebody will get an error, just open a new problem with a minimal example, thanks

**Comment by SebastienGllmt** [comment]
I spent like an hour trying to figure out why Terser refused to run in parallel no matter what I did only to realize that parallel is disabled if you are on WSL. I'm currently on Windows build 19041 (Slow ring insider preview) using WSL2 and it works perfectly with all cores.

**Comment by jerrodrs** [comment]
Upgraded my terser plugin from 1.3.5 to 2.3.5 and the issue is resolved, can use parallel now.
